Avoid allocations in NetworkedAnimator

angusmf opened the pull request:
Replicate HLAPI patch in https://bitbucket.org/Unity-Technologies/networking/pull-requests/21/cache-animator-parameters-to-avoid/diff

TwoTenPvP commented:
This doesn't work. Animator parameters is not populated before runtime. You probably have to do a check if the array is null, if so, populate it with the parameters.

angusmf commented:
Correct. Turns out you also have to remember to null the array when changing ownership. See update (and thanks for that virtual!)

TwoTenPvP commented:
Why are you resetting the array when changing ownership?

angusmf commented:
Just doing a null check wasn't sufficient when testing on HLAPI. It would get into the for loop, then get an index out of range when actually reading one of the parameters, which in some way makes sense if you've switched from serializing to deserializing or vice versa. Where I was previously initializing the parameters array I added a null check, but that left one out of range error when this happens on ownership change. So I moved it into the ResetParameterOptions and call it when a change happens. I added it to both overrides not based on testing per se. In HLAPI, they reset last sync time when auth is started, but not when it is stopped, so I observed errors when interpolation was used. I don't know that we have the same issue here, but seemed safest to just do both.

TwoTenPvP commented:
Is this relevant to the MLAPI? MLAPI has none of the connectionReady and observer issues that the HLAPI has. The reset seems unnecessary. Please try without the reset and get back to me.

angusmf commented:
I would, but don't have a test project based on MLAPI yet.

TwoTenPvP commented:
Alright, I'll give it a shot.

TwoTenPvP commented:
From my very quick testing, it's not needed. This will simply make the NetworkedAnimator nonfunctional as the parameters set to be synced are all turned off. Take it out of the PR and create another issue if neccecary.  After this merge I'll rewrite the Animator to follow the rest of the projects code standard and be a bit cleaner.

angusmf commented:
Thanks. I'll finish my test project ASAP.
